src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,311 @@
+type EntityPool<'Key, 'Value>(totalCapacity, cacheId) =
+    let pool = ConcurrentBag<CachedEntity<'Key, 'Value>>()
+    let mutable created = 0
+
+    let overCapacity =
+        CacheMetrics.Meter.CreateCounter<int64>("over-capacity", "count", cacheId)
+
+    member _.Acquire(key, value) =
+        match pool.TryTake() with
+        | true, entity -> entity.ReUse(key, value)
+        | _ ->
+            if Interlocked.Increment &created > totalCapacity then
+                overCapacity.Add 1L
+
+            // Associate a LinkedListNode with freshly created entity.
+            // This is a one time initialization.
+            CachedEntity(key, value).WithNode()
+
+    member _.Reclaim(entity: CachedEntity<'Key, 'Value>) =
+        if pool.Count < totalCapacity then
+            pool.Add(entity)
+
+module Cache =
+    // During testing a lot of compilations are started in app domains and subprocesses.
+    // This is a reliable way to pass the override to all of them.
+    [<Literal>]
+    let private overrideVariable = "FSHARP_CACHE_OVERRIDE"
+
+    /// Use for testing purposes to reduce memory consumption in testhost and its subprocesses.
+    let OverrideCapacityForTesting () =
+        Environment.SetEnvironmentVariable(overrideVariable, "true", EnvironmentVariableTarget.Process)
+
+    let applyOverride (capacity: int) =
+        match Environment.GetEnvironmentVariable(overrideVariable) with
+        | NonNull _ when capacity > 1024 -> 1024
+        | _ -> capacity
+
+[<Struct>]
+type EvictionQueueMessage<'Key, 'Value> =
+    | Add of CachedEntity<'Key, 'Value>
+    | Update of CachedEntity<'Key, 'Value>
+
+[<Sealed; NoComparison; NoEquality>]
+[<DebuggerDisplay("{GetStats()}")>]
+type Cache<'Key, 'Value when 'Key: not null and 'Key: equality> internal (totalCapacity, headroom, ?name, ?observeMetrics) =
+
+    let instanceId = defaultArg name (Guid.NewGuid().ToString())
+
+    let observeMetrics = defaultArg observeMetrics false
+
+    do
+        if observeMetrics then
+            CacheMetrics.AddInstrumentation instanceId
+
+    let meter = CacheMetrics.Meter
+    let hits = meter.CreateCounter<int64>("hits", "count", instanceId)
+    let misses = meter.CreateCounter<int64>("misses", "count", instanceId)
+    let evictions = meter.CreateCounter<int64>("evictions", "count", instanceId)
+
+    let evictionFails =
+        meter.CreateCounter<int64>("eviction-fails", "count", instanceId)
+
+    let pool = EntityPool<'Key, 'Value>(totalCapacity, instanceId)
+
+    let store =
+        ConcurrentDictionary<'Key, CachedEntity<'Key, 'Value>>(Environment.ProcessorCount, totalCapacity)
+
+    let evictionQueue = LinkedList<CachedEntity<'Key, 'Value>>()
+
+    // Non-evictable capacity.
+    let capacity = totalCapacity - headroom
+
+    let evicted = Event<_>()
+
+    let evictionProcessor =
+        new MailboxProcessor<EvictionQueueMessage<_, _>>(fun mb ->
```

Comment:
I like how this encapsulates eviction management as well as background processing now!

Reply:
Yes, I love it, too. The trade off is: now additions (and evictions) to the eviction queue are posted on the threadpool and not in sync with the store. That means the eviction queue length may lag more significantly behind the real count of items.

Reply:
Yes, I love it, too. The trade off is: now additions (and evictions) to the eviction queue are posted on the threadpool and not in sync with the store. That means the eviction queue length may lag more significantly behind the real count of items.